cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/PluginRepository.java

```diff
@@ -0,0 +1,479 @@
+      return new PluginFile(file, new PluginInfo(file.getName(), pluginName, version));
+    }
+  };
+
+  private final CConfiguration cConf;
+  private final File templateDir;
+  private final File tmpDir;
+  private final AtomicReference<Map<String, Multimap<PluginInfo, PluginClass>>> plugins;
+
+  @Inject
+  PluginRepository(CConfiguration cConf) {
+    this.cConf = cConf;
+    this.templateDir = new File(cConf.get(Constants.AppFabric.APP_TEMPLATE_DIR));
+    this.plugins = new AtomicReference<Map<String, Multimap<PluginInfo, PluginClass>>>(
+      new HashMap<String, Multimap<PluginInfo, PluginClass>>());
+    this.tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+  }
+
+  /**
+   * Returns a {@link Multimap} of all plugins available for the given application template.
+   */
+  public Multimap<PluginInfo, PluginClass> getPlugins(String template) {
+    Multimap<PluginInfo, PluginClass> result = plugins.get().get(template);
+    return result == null ? ImmutableMultimap.<PluginInfo, PluginClass>of() : result;
+  }
+
+  /**
+   * Inspects plugins for all the templates.
+   *
+   * @param templates map from template name to template jar file
+   */
+  void inspectPlugins(Map<String, File> templates) throws IOException {
+    Map<String, Multimap<PluginInfo, PluginClass>> result = Maps.newHashMap();
+    for (Map.Entry<String, File> entry : templates.entrySet()) {
+      result.put(entry.getKey(), inspectPlugins(entry.getKey(), entry.getValue()));
+    }
+
+    plugins.set(result);
+  }
+
+  /**
+   * Inspects and builds plugin information for the given application template.
+   *
+   * @param template name of the template
+   * @param templateJar application jar for the application template
+   */
+  private Multimap<PluginInfo, PluginClass> inspectPlugins(String template, File templateJar) throws IOException {
+    // We want the plugins sorted by the PluginInfo, which in turn is sorted by name and version.
+    Multimap<PluginInfo, PluginClass> templatePlugins = Multimaps.newMultimap(
+      Maps.<PluginInfo, Collection<PluginClass>>newTreeMap(), new Supplier<Collection<PluginClass>>() {
+        @Override
+        public Collection<PluginClass> get() {
+          return Lists.newArrayList();
+        }
+      });
+
+    File pluginDir = new File(templateDir, template);
+    List<File> pluginJars = DirUtils.listFiles(pluginDir, "jar");
+    if (pluginJars.isEmpty()) {
+      return templatePlugins;
+    }
+
+    Iterable<PluginFile> pluginFiles = Iterables.transform(pluginJars, FILE_TO_PLUGIN_FILE);
+    CloseableClassLoader templateClassLoader = createTemplateClassLoader(templateJar);
+    try {
+      PluginInstantiator pluginInstantiator = new PluginInstantiator(cConf, template, templateClassLoader);
+      try {
+        for (PluginFile pluginFile : pluginFiles) {
+          if (!configureByFile(pluginFile, templatePlugins)) {
```

Comment:
So a plugin configured either by file or by inspection, right? It is not possible to add a file to add additional properties for plugin that is already configured by inspection. @nitinmotgi please confirm that you agree with that